### PR TITLE
Check git version and compatibility front/backend version

### LINKED ISSUE #635: Check Git version
perhaps you want to check the version of git in jupyterlab-git and warn the user, if they use a too old git (i.e. git < 2)

_Originally posted by @jhgoebbert in https://github.com/jupyterlab/jupyterlab-git/issues/625#issuecomment-625235178_

### LINKED ISSUE #636: Unable to detect a Git repository.
I installed the program extention with:

```
conda create -c conda-forge -n MLPB -y python=3.8 pandas>1.0 scikit-learn pip matplotlib seaborn biopython pyvis jupyterlab jupyterlab-git ipywidgets jupyter_nbextensions_configurator xlrd statsmodels
conda activate MLPB
jupyter labextension install @jupyter-widgets/jupyterlab-manager
jupyter labextension install @jupyterlab/git
```

The extension appears in Jupyter lab but it is unable to detect existing git repositories. 

Clicking on detect repositories gets me to the file browser. I can navigate to a git repository and go back to the git tab, but the same message appears.

### PR DESCRIPTION
**Pull request by fcollonval**
Fixes #635
Closes #636

Before doing anything, the plugin does a request to the server extension to get :
- the server root path
- the git version
- the server extension version
- the frontend extension version formatted as the server one

if that request returns 404, git version < 2 or front/backend versions do not match => nothing will be loaded and an error message will be shown to the user - with enough information on what went wrong.

There is a not-to-nice part in the sense that the frontend version is sent to the backend so it gets formatted similarly through the `packaging` Python package. That package can interpret the npm prerelease version while semver can't interpret Python prerelease version - hence the design choice.

> The frontend version is automatically extract when building the package using `genversion` package.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -77,8 +80,40 @@ async function activate(
...
+    }
+
+    if (
+      serverSettings.frontendVersion &&
+      serverSettings.frontendVersion !== serverSettings.serverVersion
+    ) {
+      throw new Error(
+        'JupyterLab Git server extension and frontend extension are not matching. ' +
+          'Please install identical version of jupyterlab-git Python package and @jupyterlab/git extension.'
+      );
```
Comment (ianhi) [inline src/index.ts:105]:
```suggestion
    throw new Error(
      'The versions of the JupyterLab Git server frontend and backend do not match. ' +
        `The @jupyterlab/git frontend extension has version: ${
          serverSettings.frontendVersion
        } ` +
        `while the python package has version ${
          serverSettings.serverVersion
        } ` +
        'Please install identical version of jupyterlab-git Python package and the @jupyterlab/git extension. Try running: pip install --upgrade jupyterlab-git'
    );
```

File: jupyterlab_git/handlers.py [hunk 2]
```diff
@@ -534,13 +536,31 @@ async def post(self):
...
     @web.authenticated
     async def get(self):
+        jlab_version = self.get_query_argument("version", None)
+        if jlab_version is not None:
+            jlab_version = str(parse(jlab_version))
+        git_version = None
+        try:
+            git_version = await self.git.version()
+        except Exception as error:
+            self.log.debug("[jupyterlab_git] Fail to execute 'git' command: {!s}".format(error))
```
Comment (ianhi) [inline jupyterlab_git/handlers.py:549]:
```suggestion
            self.log.debug("[jupyterlab_git] Failed to execute 'git' command: {!s}".format(error))
```

File: src/index.ts (hunk 1 above)
Comment (ianhi) [reply src/index.ts:105]: Could be nice to give more information about what went wrong and how to fix it.

File: src/index.ts [hunk 3]
```diff
@@ -77,8 +80,40 @@ async function activate(
...
+    }
+  } catch (error) {
+    // If we fall here, nothing will be loaded in the frontend.
+    console.error('Fail to load the server git extension settings', error);
+    showErrorMessage('Fail to load the server git extension', error.message, [
```
Comment (ianhi) [inline src/index.ts:110]:
```suggestion
    console.error('Failed to load the jupyterlab-git server extension settings', error);
    showErrorMessage('Failed to load the jupyterlab-git server extension', error.message, [
```

File: src/index.ts [hunk 4]
```diff
@@ -77,8 +80,40 @@ async function activate(
   } catch (error) {
     console.error(`Failed to load settings for the Git Extension.\n${error}`);
   }
+  try {
+    serverSettings = await getServerSettings();
+
+    // Version validation
+    if (!serverSettings.gitVersion) {
+      throw new Error('git command not found');
```
Comment (ianhi) [inline src/index.ts:88]:
```suggestion
      throw new Error('Git command not found - please ensure you have Git > 2 installed');
```

File: tests/plugin.spec.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        null,
+        { defaultBrowser: null },
+        null,
+        settingRegistry
+      );
+
+      // Then
+      expect(extension).toBeNull(); // Token is null
+      expect(mockedErrorMessage).toHaveBeenCalledWith(
+        'Fail to load the server git extension',
```
Comment (ianhi) [inline tests/plugin.spec.ts:193]:
```suggestion
        'Failed to load the jupyterlab-git server extension',
```

File: tests/plugin.spec.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        'JupyterLab Git server extension and frontend extension are not matching. ' +
+          'Please install identical version of jupyterlab-git Python package and @jupyterlab/git extension.',
```
Comment (ianhi) [inline tests/plugin.spec.ts:165]:
```suggestion
      'The versions of the JupyterLab Git server frontend and backend do not match. ' +
        `The @jupyterlab/git frontend extension has version: ${
          version
        } ` +
        'while the python package has version 0.1.0' +
        'Please install identical version of jupyterlab-git Python package and the @jupyterlab/git extension. Try running: pip install --upgrade jupyterlab-git',
```

I may have goofed the linting here.

File: src/server.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        'To confirm that the server extension is installed, run: jupyter serverextension list.';
+      throw new ServerConnection.ResponseError(response, message);
+    }
+    let content: string | any = await response.text();
+    if (content.length > 0) {
+      content = JSON.parse(content);
+    }
+    if (!response.ok) {
+      const message = content.message || content;
+      console.error('Fail to get the server extension settings', message);
```
Comment (ianhi) [inline src/server.ts:24]:
```suggestion
      console.error('Failed to get the server extension settings', message);
```

File: tests/plugin.spec.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
+import 'jest';
+import * as git from '../src/git';
+// import { Git } from '../src/tokens';
```
Comment (ianhi) [inline tests/plugin.spec.ts:3]:
```suggestion
```

File: tests/plugin.spec.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+      // Then
+      expect(extension).toBeNull(); // Token is null
+      expect(mockedErrorMessage).toHaveBeenCalledWith(
+        'Fail to load the server git extension',
+        'git command version must be > 2; got 1.8.7.',
+        [undefined] // The warning button is undefined as the module @jupyterlab/apputils is mocked
+      );
+    });
+    it('should fail if server and extension version are not matching', async () => {
```
Comment (ianhi) [inline tests/plugin.spec.ts:132]:
```suggestion
    it('should fail if server and extension version do not match', async () => {
```

File: tests/plugin.spec.ts [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        null,
+        { defaultBrowser: null },
+        null,
+        settingRegistry
+      );
+
+      // Then
+      expect(extension).toBeNull(); // Token is null
+      expect(mockedErrorMessage).toHaveBeenCalledWith(
+        'Fail to load the server git extension',
```
Comment (ianhi) [inline tests/plugin.spec.ts:163]:
```suggestion
        'Failed to load the jupyterlab-git server extension',
```

**Comment by fcollonval** [comment]
Thanks for the review @ianhi

File: src/index.ts (hunk 4 above)
Comment (fcollonval) [reply src/index.ts:88]: I updated the suggestion to keep `git` (lowercase) to be consistent with the command name.

**Comment by fcollonval** [comment]
@all-contributors please add @ianhi for review

**Comment by fcollonval** [comment]
@meeseeksdev backport to 0.11.x

**Pull request merged**